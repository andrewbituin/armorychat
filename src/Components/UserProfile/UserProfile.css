.user-profile-container {
  padding: 0 20px;
}

.user-profile__input-field {
  display: flex;
  justify-content: space-around;
  list-style: none;
  padding: 0;
  flex-wrap: wrap;
}

.user-profile__avatar-container, .user-profile__avatar-fake-checkbox {
  position: relative;
  display: inline-flex;
  justify-content: center;
  align-items: center;
  width: 70px;
  height: 70px;
  overflow: hidden;
}

.user-profile__avatar-container:focus {
  outline: none;
}

.user-profile__avatar-container:hover {
  cursor: pointer;
}

.user-profile__avatar-icon-container {
  position: relative;
  box-sizing: content-box;
  width: 50px;
  height: 50px;
  margin: 2px;
  border-radius: 5px;
  box-shadow: 0 1px 8px 0 #00000080;
  border: 1px solid #909193;
  overflow: hidden;
}

.user-profile__avatar-fake-image {
  position: absolute;
  width: 50px;
}

.user-profile__avatar-image {
  box-sizing: content-box;
  width: 50px;
  margin: 2px;
  border-radius: 5px;
  box-shadow: 0 1px 8px 0 #00000080;
  border: 1px solid #909193;
}

.user-profile__avatar-image:hover {
  cursor: pointer;
}

.user-profile__avatar-image:hover:not(.user-profile__avatar-image-active) {
  border: 3px solid #5a801f;
}

.user-profile__avatar-image-active{
  border: 4px solid #5a801f;
}

.flex-fieldset-column {
  display: flex;
  flex-direction: column;
}

.user-profile__submit-button, .user-profile__cancel-button {
  height: 30px;
}

.user-profile__submit-button {
  width: 60%;
}

.user-profile__cancel-button {
  width: 25%;
}

@media screen and (max-width: 500px) {
  .user-profile__avatar-container, .user-profile__avatar-fake-checkbox {
    width: 50px;
  }

  .user-profile__avatar-icon-container, 
  .user-profile__avatar-fake-image,
  .user-profile__avatar-image {
    width: 40px;
    height: 40px;
  }

}

