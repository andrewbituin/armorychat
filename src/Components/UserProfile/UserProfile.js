import React, { useState, useContext, useEffect } from "react";
import UserContext from "../../Contexts/userContext";
import TokenService from "../../services/token-service";
import config from '../../config';
import AuthApiService from "../../services/auth-api-service";
import './UserProfile.css';

export default function UserProfile(props) {
  const [email, setEmail] = useState("");
  const [avatar_url, setAvatar_url] = useState("");
  const [currentPass, setCurrentPass] = useState("");
  const [newPass, setNewPass] = useState("");
  const [confirmPass, setConfirmPass] = useState("");
  const [error, setError] = useState("");
  const context = useContext(UserContext);
  const curr = context.user;
  const currentEmail = curr.email;

  useEffect(() => {
    populateContext().then(user => {
      context.setUser(user.userInfo);
    });
  }, []);

  function populateContext() {
    const { user_id } = TokenService.parseAuthToken();
    return fetch(`${config.API_ENDPOINT}/user/${user_id}`, {
      method: "GET",
      headers: {
        authorization: `Bearer ${TokenService.getAuthToken()}`
      }
    }).then(res => {
      return !res.ok ? TokenService.clearAuthToken() : res.json();
    });
  }

  // console.log(props.update)
  
  function confirmNewPassword(newPass, confirmPass) {
    if (newPass) {
      if (newPass === confirmPass) {
        let user = { email, avatar_url, password: newPass };
        saveChanges(user);
        setNewPass("");
        setConfirmPass("");
        setCurrentPass("");
      } else {
        setError("Passwords do not match");
        return;
      }
    } else {
      let user = { email, avatar_url };
      saveChanges(user);
      setEmail('');
    }
  }
  
  function saveChanges(user) {
    const { user_id } = TokenService.parseAuthToken();
    return fetch(`${config.API_ENDPOINT}/user/${user_id}`, {
      method: "PATCH",
      headers: {
        "content-type": "application/json",
        authorization: `Bearer ${TokenService.getAuthToken()}`
      },
      body: JSON.stringify(user)
    })
      .then(res =>
        !res.ok ? setError(res.error) : props.update() && props.toggle()
      )
      .then(props.toggle);
  }

  function authorizeChanges(e) {
    e.preventDefault();
    AuthApiService.postLogin({ email: currentEmail, password: currentPass })
      .then(res => {
        // console.log(res.authToken);
        TokenService.saveAuthToken(res.authToken);
        confirmNewPassword(newPass, confirmPass);
      })
      .catch(error => setError(error.error));
  }

  function generateUserIconImages() {
    const imageNames = ['Default-Avatar.png', 'ninja.png', 'pikachu.png', 'soldier76.png', 'cactuar.png', 'axe.png', 'TJICON.png'];
    
    return imageNames.map((name, i) => {
      if (context.user.avatar_url === name) {
        return (
          <li 
            key={i}
            className="user-profile__avatar-fake-checkbox"
          >
            <div className="user-profile__avatar-icon-container">
              <img
                className="user-profile__avatar-fake-image"
                src={`${config.IMAGES_ENDPOINT}/user-icons/${name}`}
                alt={name}
              />
              <i className="squad__spots-filled fas fa-check"/>
            </div>
          </li>
        )
      } else {
        return (
          <li
            aria-label="checkbox"
            tabIndex="0"
            onClick={e => {
              if (avatar_url === name) {
                setAvatar_url('');
              } else {
                setAvatar_url(name);
              }
            }}
            className="user-profile__avatar-container"
            key={i}
          >
            <img
              src={`${config.IMAGES_ENDPOINT}/user-icons/${name}`}
              alt={name}
              className={avatar_url === name ? "user-profile__avatar-image user-profile__avatar-image-active" : "user-profile__avatar-image"}
            />
          </li>
        )
      }
    });
  }

  return (
    <div className="user-profile-container">
      <h2 className="modal-header">Profile Settings</h2>
      <form className="profile-form" onSubmit={e => authorizeChanges(e)}>
        <p>{error}</p>

        <fieldset className="create-squad__fieldset">
          <legend>Choose a new avatar:</legend>
          <ul className="user-profile__input-field">
            {generateUserIconImages()}
          </ul>
        </fieldset>

        <fieldset className="create-squad__fieldset">
          <legend>Enter a new email:</legend>
          <input
            aria-label="profile-email-input"
            type="email"
            placeholder={curr.email}
            className="create-squad__name-input"
            value={email}
            onChange={e => setEmail(e.target.value)}
          />
        </fieldset>

        <fieldset className="flex-fieldset-column create-squad__fieldset">
          <legend>Enter a new password:</legend>

          <input
            aria-label="profile-new-password-input"
            type="password"
            placeholder="New password..."
            className="create-squad__name-input"
            value={newPass}
            onChange={e => setNewPass(e.target.value)}
          />

          <input
            aria-label="profile-new-c-password-input"
            type="password"
            placeholder="Confirm new password..."
            className="create-squad__name-input"
            value={confirmPass}
            onChange={e => setConfirmPass(e.target.value)}
          />
        </fieldset>
       <fieldset className="flex-fieldset-column create-squad__fieldset">
          <legend>Confirm your password:</legend>
          <input
            aria-label="profile-password-input"
            type="password"
            placeholder="Current password..."
            className="create-squad__name-input"
            required
            value={currentPass}
            onChange={e => setCurrentPass(e.target.value)}
          />
        </fieldset>

        <div className="modal-button-holder create-squad__button-container">
          <button type="submit" className="user-profile__submit-button green-button">
            Update Profile
          </button>
          <button type="button" aria-label="cancel" className="user-profile__cancel-button grey-button" onClick={props.toggle}>
            Cancel
          </button>
        </div>
      </form>
    </div>
  );
}
