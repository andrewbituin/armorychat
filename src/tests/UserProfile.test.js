import React from 'react';
import UserProfile from '../Components/UserProfile/UserProfile';
import { MemoryRouter } from 'react-router-dom';
import renderer from "react-test-renderer";
import { shallow } from "enzyme";

describe('UserProfile component', () => {
  it('renders without crashing', () => {
    shallow(<UserProfile/>)
  })

  it('when user profile form is submitted the event is cancelled', () => {
    const wrapper = shallow(<UserProfile />);
    let prevented = false;
    wrapper.find('form').simulate('submit', {
      preventDefault: () => {
        prevented = true;
      }
    });
    expect(prevented).toBe(true);
  })

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <MemoryRouter>
          <UserProfile />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})