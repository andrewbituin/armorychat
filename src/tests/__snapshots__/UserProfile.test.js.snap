// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserProfile component renders the UI as expected 1`] = `
<div
  className="user-profile-container"
>
  <h2
    className="modal-header"
  >
    Profile Settings
  </h2>
  <form
    className="profile-form"
    onSubmit={[Function]}
  >
    <p>
      
    </p>
    <fieldset
      className="create-squad__fieldset"
    >
      <legend>
        Choose a new avatar:
      </legend>
      <ul
        className="user-profile__input-field"
      >
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="Default-Avatar.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/Default-Avatar.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="ninja.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/ninja.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="pikachu.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/pikachu.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="soldier76.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/soldier76.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="cactuar.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/cactuar.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="axe.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/axe.png"
          />
        </li>
        <li
          aria-label="checkbox"
          className="user-profile__avatar-container"
          onClick={[Function]}
          tabIndex="0"
        >
          <img
            alt="TJICON.png"
            className="user-profile__avatar-image"
            src="https://limitless-brushlands-45977.herokuapp.com/images/user-icons/TJICON.png"
          />
        </li>
      </ul>
    </fieldset>
    <fieldset
      className="create-squad__fieldset"
    >
      <legend>
        Enter a new email:
      </legend>
      <input
        aria-label="profile-email-input"
        className="create-squad__name-input"
        onChange={[Function]}
        type="email"
        value=""
      />
    </fieldset>
    <fieldset
      className="flex-fieldset-column create-squad__fieldset"
    >
      <legend>
        Enter a new password:
      </legend>
      <input
        aria-label="profile-new-password-input"
        className="create-squad__name-input"
        onChange={[Function]}
        placeholder="New password..."
        type="password"
        value=""
      />
      <input
        aria-label="profile-new-c-password-input"
        className="create-squad__name-input"
        onChange={[Function]}
        placeholder="Confirm new password..."
        type="password"
        value=""
      />
    </fieldset>
    <fieldset
      className="flex-fieldset-column create-squad__fieldset"
    >
      <legend>
        Confirm your password:
      </legend>
      <input
        aria-label="profile-password-input"
        className="create-squad__name-input"
        onChange={[Function]}
        placeholder="Current password..."
        required={true}
        type="password"
        value=""
      />
    </fieldset>
    <div
      className="modal-button-holder create-squad__button-container"
    >
      <button
        className="user-profile__submit-button green-button"
        type="submit"
      >
        Update Profile
      </button>
      <button
        aria-label="cancel"
        className="user-profile__cancel-button grey-button"
        type="button"
      >
        Cancel
      </button>
    </div>
  </form>
</div>
`;
